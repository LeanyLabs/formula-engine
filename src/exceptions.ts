export class FormulaError extends Error {
  public details: any;
  constructor(message: string, details: any) {
    super(message);
    this.details = details;
  }
}

export class ParserError extends FormulaError {}
export class LexerError extends FormulaError {}

export class ExecutionError extends FormulaError {}
export class FunctionError extends ExecutionError {}
