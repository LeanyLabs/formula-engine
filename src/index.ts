import {
  FormulaEngine,
  ReferenceResolver,
  ReferenceResolverAsync,
} from "./engine";

export { FormulaEngine, ReferenceResolver, ReferenceResolverAsync };
