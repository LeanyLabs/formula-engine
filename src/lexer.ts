import { createToken, Lexer, TokenType } from "chevrotain";

enum TokenName {
  AdditionOperator = "AdditionOperator",
  Plus = "Plus",
  Minus = "Minus",
  MultiplicationOperator = "MultiplicationOperator",
  Multi = "Multi",
  Div = "Div",

  LParen = "LParen",
  RParen = "RParen",

  WhiteSpace = "WhiteSpace",
  Comma = "Comma",

  Function = "Function",
  Reference = "Reference",

  NumberLiteral = "NumberLiteral",
  DStringLiteral = "DStringLiteral",
  SStringLiteral = "SStringLiteral",
  BooleanLiteral = "BooleanLiteral",

  ComparisonOperator = "ComparisonOperator",
  Lt = "Lt",
  Gt = "Gt",
  Lte = "Lte",
  Gte = "Gte",
  Eq = "Eq",
}

const AdditionOperator = createToken({
  name: TokenName.AdditionOperator,
  pattern: Lexer.NA,
});
const Plus = createToken({
  name: TokenName.Plus,
  pattern: /\+/,
  categories: AdditionOperator,
});
const Minus = createToken({
  name: TokenName.Minus,
  pattern: /-/,
  categories: AdditionOperator,
});

const MultiplicationOperator = createToken({
  name: TokenName.MultiplicationOperator,
  pattern: Lexer.NA,
});
const Multi = createToken({
  name: TokenName.Multi,
  pattern: /\*/,
  categories: MultiplicationOperator,
});
const Div = createToken({
  name: TokenName.Div,
  pattern: /\//,
  categories: MultiplicationOperator,
});

const LParen = createToken({
  name: TokenName.LParen,
  pattern: /\(/,
});
const RParen = createToken({
  name: TokenName.RParen,
  pattern: /\)/,
});

const Function = createToken({
  name: TokenName.Function,
  pattern: /[A-Za-z_]+[A-Za-z_0-9.]*/,
});
const Reference = createToken({
  name: TokenName.Reference,
  pattern: /\{[A-Za-z_0-9\.]+\}/,
});

const Comma = createToken({
  name: TokenName.Comma,
  pattern: /,/,
});
const WhiteSpace = createToken({
  name: "WhiteSpace",
  pattern: /\s+/,
  group: Lexer.SKIPPED,
});

const BooleanLiteral = createToken({
  name: TokenName.BooleanLiteral,
  pattern: /TRUE|FALSE/i,
});
const NumberLiteral = createToken({
  name: TokenName.NumberLiteral,
  pattern: /[0-9]+[.]?[0-9]*([eE][+\-][0-9]+)?/,
});
const DStringLiteral = createToken({
  name: TokenName.DStringLiteral,
  pattern: /"(""|[^"])*"/,
});
const SStringLiteral = createToken({
  name: TokenName.SStringLiteral,
  pattern: /'(''|[^'])*'/,
});

const ComparisonOperator = createToken({
  name: TokenName.ComparisonOperator,
  pattern: Lexer.NA,
});
const Lt = createToken({
  name: TokenName.Lt,
  pattern: /</,
  categories: ComparisonOperator,
});
const Gt = createToken({
  name: TokenName.Gt,
  pattern: />/,
  categories: ComparisonOperator,
});
const Lte = createToken({
  name: TokenName.Lte,
  pattern: /<=/,
  categories: ComparisonOperator,
});
const Gte = createToken({
  name: TokenName.Gte,
  pattern: />=/,
  categories: ComparisonOperator,
});
const Eq = createToken({
  name: TokenName.Eq,
  pattern: /==/,
  categories: ComparisonOperator,
});

const tokensByPriority = [
  WhiteSpace,
  Plus,
  Minus,
  Multi,
  Div,
  Lte,
  Lt,
  Gte,
  Gt,
  Eq,
  LParen,
  RParen,
  DStringLiteral,
  SStringLiteral,
  BooleanLiteral,
  NumberLiteral,
  ComparisonOperator,
  AdditionOperator,
  MultiplicationOperator,
  Function,
  Reference,
  Comma,
];

export const FormulaLexer = new Lexer(tokensByPriority, {
  ensureOptimizations: true,
});

export type TokenTypeDict = { [key in TokenName]: TokenType };
export const tokens: TokenTypeDict = tokensByPriority.reduce(
  (acc, tokenType) => {
    acc[tokenType.name] = tokenType;
    return acc;
  },
  {} as TokenTypeDict
);
