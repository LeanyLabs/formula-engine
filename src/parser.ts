import { CstParser } from "chevrotain";
import { tokens } from "./lexer";

export class FormulaParser extends CstParser {
  constructor() {
    super(tokens, {
      maxLookahead: 1,
    });
    this.performSelfAnalysis();
  }

  expression = this.RULE("expression", () => {
    this.SUBRULE(this.comparisonExpression);
  });

  comparisonExpression = this.RULE("comparisonExpression", () => {
    this.SUBRULE(this.additionExpression, { LABEL: "lhs" });
    this.OPTION(() => {
      this.CONSUME(tokens.ComparisonOperator);
      this.SUBRULE1(this.additionExpression, { LABEL: "rhs" });
    });
  });

  additionExpression = this.RULE("additionExpression", () => {
    this.SUBRULE(this.multiplicationExpression, { LABEL: "lhs" });
    this.MANY(() => {
      this.CONSUME(tokens.AdditionOperator);
      this.SUBRULE1(this.multiplicationExpression, { LABEL: "rhs" });
    });
  });

  multiplicationExpression = this.RULE("multiplicationExpression", () => {
    this.SUBRULE(this.atomicExpression, { LABEL: "lhs" });
    this.MANY(() => {
      this.CONSUME(tokens.MultiplicationOperator);
      this.SUBRULE1(this.atomicExpression, { LABEL: "rhs" });
    });
  });

  atomicExpression = this.RULE("atomicExpression", () => {
    this.OR([
      { ALT: () => this.SUBRULE(this.functionExpression) },
      { ALT: () => this.SUBRULE(this.parenthesisExpression) },
      { ALT: () => this.SUBRULE(this.reference) },
      { ALT: () => this.CONSUME(tokens.NumberLiteral) },
      { ALT: () => this.CONSUME(tokens.BooleanLiteral) },
      { ALT: () => this.CONSUME(tokens.SStringLiteral) },
      { ALT: () => this.CONSUME(tokens.DStringLiteral) },
    ]);
  });

  reference = this.RULE("reference", () => {
    this.CONSUME(tokens.Reference);
  });

  functionExpression = this.RULE("functionExpression", () => {
    this.CONSUME(tokens.Function);
    this.CONSUME(tokens.LParen);
    this.SUBRULE(this.arguments);
    this.CONSUME(tokens.RParen);
  });

  parenthesisExpression = this.RULE("parenthesisExpression", () => {
    this.CONSUME(tokens.LParen);
    this.SUBRULE(this.expression);
    this.CONSUME(tokens.RParen);
  });

  arguments = this.RULE("arguments", () => {
    this.MANY_SEP({
      SEP: tokens.Comma,
      DEF: () => {
        this.SUBRULE(this.additionExpression);
      },
    });
  });
}
