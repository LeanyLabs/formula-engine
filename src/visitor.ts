import { CstNode, tokenMatcher } from "chevrotain";
import { ExecutionError, FunctionError } from "./exceptions";
import { tokens } from "./lexer";
import { FormulaParser } from "./parser";

export interface IVisitor {
  visit(cst: CstNode, state: any): any;
}

export function createEvalVisitor(
  parser: FormulaParser,
  functions: any
): IVisitor {
  const FormulaVisitorBase = parser.getBaseCstVisitorConstructorWithDefaults();

  class InterpreterVisitor extends FormulaVisitorBase {
    constructor() {
      super();
      this.validateVisitor();
    }

    expression(ctx, state): any {
      return this.visit(ctx.comparisonExpression, state);
    }

    comparisonExpression(ctx, state): any {
      let left = this.visit(ctx.lhs, state);
      if (!ctx.rhs) return left;

      let right = this.visit(ctx.rhs, state);
      const operator = ctx.ComparisonOperator[0];

      if (tokenMatcher(operator, tokens.Lt)) {
        return left < right;
      }
      if (tokenMatcher(operator, tokens.Lte)) {
        return left <= right;
      }
      if (tokenMatcher(operator, tokens.Gt)) {
        return left > right;
      }
      if (tokenMatcher(operator, tokens.Gte)) {
        return left >= right;
      }
      if (tokenMatcher(operator, tokens.Eq)) {
        return left === right;
      }
      throw new ExecutionError(`Operator not implemented: ${operator}`, {
        operator,
        context: ctx,
      });
    }

    additionExpression(ctx, state): any {
      let result = this.visit(ctx.lhs, state);
      if (!ctx.rhs) return result;
      for (let i = 0; i < ctx.rhs.length; i++) {
        const operator = ctx.AdditionOperator[i];
        const value = this.visit(ctx.rhs[i], state);
        if (tokenMatcher(operator, tokens.Plus)) {
          result += value;
        } else if (tokenMatcher(operator, tokens.Minus)) {
          result -= value;
        } else {
          throw new ExecutionError(
            `Unknown operator: ${operator.image} at ${operator.startOffset}`,
            { operator, context: ctx }
          );
        }
      }
      return result;
    }

    multiplicationExpression(ctx, state): any {
      let result = this.visit(ctx.lhs, state);
      if (!ctx.rhs) return result;
      for (let i = 0; i < ctx.rhs.length; i++) {
        const operator = ctx.MultiplicationOperator[i];
        const value = this.visit(ctx.rhs[i], state);
        if (tokenMatcher(operator, tokens.Multi)) {
          result *= value;
        } else if (tokenMatcher(operator, tokens.Div)) {
          result /= value;
        } else {
          throw new ExecutionError(
            `Unknown operator: ${operator.image} at ${operator.startOffset}`,
            { operator, context: ctx }
          );
        }
      }
      return result;
    }

    atomicExpression(ctx, state): any {
      if (ctx.functionExpression) {
        return this.visit(ctx.functionExpression, state);
      }
      if (ctx.parenthesisExpression) {
        return this.visit(ctx.parenthesisExpression, state);
      }
      if (ctx.reference) {
        return this.visit(ctx.reference, state);
      }
      if (ctx.NumberLiteral) {
        return Number.parseFloat(ctx.NumberLiteral[0].image);
      }
      if (ctx.BooleanLiteral) {
        return ctx.BooleanLiteral[0].image === "true" ? true : false;
      }
      if (ctx.SStringLiteral) {
        return ctx.SStringLiteral[0].image.slice(1, -1).replace(/''/, "'");
      }
      if (ctx.DStringLiteral) {
        return ctx.DStringLiteral[0].image.slice(1, -1).replace(/""/, '"');
      }
    }

    functionExpression(ctx, state): any {
      const functionName = ctx.Function[0].image;
      const func = functions[functionName];

      if (!func) {
        throw new ExecutionError(
          `Unknown function: ${functionName} at ${ctx.Function[0].startOffset}`,
          { functionName, context: ctx }
        );
      }

      const args = this.visit(ctx.arguments[0], state);

      try {
        const result = func.apply({}, args);
        return result;
      } catch (err) {
        throw new FunctionError(
          `Function ${functionName} at ${ctx.Function[0].startOffset} thrown an error: ${err}, stacktrace: ${err.stack}`,
          {
            originalError: err,
            functionName,
            function: ctx.Function[0],
            context: ctx,
          }
        );
      }
    }

    reference(ctx, state): any {
      const name = ctx.Reference[0].image.slice(1, -1);
      return state.externals[name];
    }

    parenthesisExpression(ctx, state): any {
      return this.visit(ctx.expression, state);
    }

    arguments(ctx, state): any {
      if (!ctx.additionExpression) return [];
      const result = ctx.additionExpression.map((arg) =>
        this.visit(arg, state)
      );
      return result;
    }
  }

  return new InterpreterVisitor();
}

export function createRefVisitor(parser: FormulaParser): IVisitor {
  const FormulaVisitorBase = parser.getBaseCstVisitorConstructorWithDefaults();

  class RefVisitor extends FormulaVisitorBase {
    constructor() {
      super();
      this.validateVisitor();
    }

    reference(ctx, state): any {
      const name = ctx.Reference[0].image.slice(1, -1);
      return state.externalNames.push(name);
    }
  }

  return new RefVisitor();
}
