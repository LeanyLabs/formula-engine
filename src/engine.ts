import { LexerError, ParserError } from "./exceptions";
import { FormulaLexer } from "./lexer";
import { FormulaParser } from "./parser";
import { createEvalVisitor, createRefVisitor, IVisitor } from "./visitor";

export type FunctionsResolver = Record<string, Function>;
export type ReferenceResolver = (names: string[]) => Record<string, any>;
export type ReferenceResolverAsync = (
  names: string[]
) => Promise<Record<string, any>>;

export class FormulaEngine {
  private parser: FormulaParser;
  private evalVisitor: IVisitor;
  private refVisitor: IVisitor;

  constructor(
    private resolveReferences: ReferenceResolver | ReferenceResolverAsync,
    private functions: FunctionsResolver
  ) {
    this.parser = new FormulaParser();
    this.evalVisitor = createEvalVisitor(this.parser, this.functions);
    this.refVisitor = createRefVisitor(this.parser);
  }

  exec(formula: string) {
    const cst = this.getCst(formula);

    const externalNames = [];
    this.refVisitor.visit(cst, { externalNames });
    const externals = this.resolveReferences(externalNames);

    const result = this.evalVisitor.visit(cst, { externals });
    return result;
  }

  async execAsync(formula: string) {
    const cst = this.getCst(formula);

    const externalNames = [];
    this.refVisitor.visit(cst, { externalNames });
    const externals = await (this.resolveReferences as ReferenceResolverAsync)(
      externalNames
    );

    const result = this.evalVisitor.visit(cst, { externals });
    return result;
  }

  private getCst(formula: string) {
    const lexingResult = this.tokenize(formula);
    const cst = this.parse(lexingResult);
    return cst;
  }

  private tokenize(formula: string) {
    const lexingResult = FormulaLexer.tokenize(formula);
    if (lexingResult.errors.length > 0) {
      throw new LexerError("Tokenization error", lexingResult.errors);
    }
    return lexingResult;
  }

  private parse(lexingResult) {
    this.parser.reset();
    this.parser.input = lexingResult.tokens;
    const cst = this.parser.expression();

    if (!cst || this.parser.errors.length > 0) {
      throw new ParserError("Parsing error", this.parser.errors);
    }

    return cst;
  }
}
