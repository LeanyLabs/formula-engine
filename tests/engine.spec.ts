import { FormulaEngine } from "../src/engine";
import * as functions from "./test-functions";

describe("Engine", () => {
  function resolveReferencesSync(names: string[]): Record<string, any> {
    return {
      a: 10,
      b: 20,
      'x.y': 5,
    };
  }
  const engine = new FormulaEngine(resolveReferencesSync, functions);

  function assertEqual(formula: string, result) {
    expect(engine.exec(formula)).toEqual(result);
  }

  describe("simple operations", () => {
    describe("add/subtract", () => {
      it("should work for literals", () => {
        assertEqual("10 + 20", 30);
      });
      it("should work for references", () => {
        assertEqual("{a} + {b}", 30);
      });
      it("should work for references with dot in the name", () => {
        assertEqual("{x.y} + {a}", 15);
      });
      it("should work for literals - sub", () => {
        assertEqual("10 - 20", -10);
      });
    });

    describe("mul/div", () => {
      it("should work for literals", () => {
        assertEqual("10 * 20", 200);
      });
      it("should work for references", () => {
        assertEqual("{a} * {b}", 200);
      });
    });

    describe("priorities", () => {
      it("should do mul before addition", () => {
        assertEqual("1 + 2 * 3", 7);
      });

      it("should do div before addition", () => {
        assertEqual("1 + 3 / 3", 2);
      });

      it("should do op in parens first", () => {
        assertEqual("(1 + 3) / 2", 2);
      });

      it("should do add/mul before comparison", () => {
        assertEqual("1 + 4 / 2 < 3 + 1", true);
      });
    });

    describe("formatting", () => {
      it("should work w/o whitespace separators", () => {
        assertEqual("10+20", 30);
        assertEqual("10+ 20", 30);
        assertEqual("10 +20", 30);
        assertEqual("1 + 2 *3", 7);
        assertEqual("1+2 *3", 7);
      });
    });

    describe("functions", () => {
      it("should use custom functions", () => {
        assertEqual("ADD(1, 2)", 3);
      });

      it("should calc function arguments", () => {
        assertEqual("ADD(1+2, 3)", 6);
      });

      it("should resolve refs", () => {
        assertEqual("ADD({a}, 3)", 13);
      });

      it("should calc functions as arguments", () => {
        assertEqual("ADD(SUB({a}, 3), 10)", 17);
      });

      it("should work with strings as return value", async () => {
        assertEqual("'The date is: ' + NOW()", "The date is: 01-02-2021");
      });
    });
  });

  describe("data types", () => {
    describe("strings", () => {
      it("should add strings", () => {
        assertEqual("'a' + 'b'", "ab");
      });

      it("should add dbl quoted strings", () => {
        assertEqual(`"a" + "b"`, "ab");
      });

      it("should be able to use string as a function argument", async () => {
        assertEqual(`TO_UPPER('abc')`, "ABC");
      });
    });
  });

  describe("comparison", () => {
    it("should compare two values", async () => {
      assertEqual(`10 > 20`, false);
      assertEqual(`10 < 20`, true);
    });

    it("should compare two values with lt/gt equal", async () => {
      assertEqual(`10 >= 20`, false);
      assertEqual(`30 <= 20`, false);

      assertEqual(`10 <= 10`, true);
      assertEqual(`10 >= 10`, true);
    });

    it("should compare two values for equality", async () => {
      assertEqual(`10 == 20`, false);
      assertEqual(`20 == 20`, true);
    });

    it("should compare string for equality", async () => {
      assertEqual(`'one' == "one"`, true);
      assertEqual(`'one' == 'one'`, true);
      assertEqual(`'one' == 'two'`, false);
      assertEqual(`'one' == 'ONE'`, false);
    });

    it("should compare variables", async () => {
      assertEqual(`{a} < {b}`, true);
      assertEqual(`{a} > {b}`, false);
      assertEqual(`{a} == 10`, true);
    });
  });

  describe("invalid syntax", () => {
    it("should throw ParsingError", async () => {
      expect(() => engine.exec(`10 <*`)).toThrowErrorMatchingSnapshot();
    });

    it("should throw LexerError", async () => {
      expect(() => engine.exec(`^`)).toThrowErrorMatchingSnapshot();
    });
  });

  describe("functions", () => {
    it("should throw FunctionError if function fails", async () => {
      expect(() => engine.exec(`THROW()`)).toThrowError();
    });
  });
});
