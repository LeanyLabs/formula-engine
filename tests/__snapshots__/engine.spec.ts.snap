// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine functions should throw FunctionError if function fails 1`] = `
"Function THROW at 0 thrown an error: Error: function failed, stacktrace: Error: function failed
    at Object.THROW (/Users/mzahor/leany/formula-engine/tests/test-functions.ts:18:9)
    at InterpreterVisitor.functionExpression (/Users/mzahor/leany/formula-engine/src/visitor.ts:126:29)
    at InterpreterVisitor.visit (/Users/mzahor/leany/formula-engine/node_modules/chevrotain/src/parse/cst/cst_visitor.ts:61:32)
    at InterpreterVisitor.atomicExpression (/Users/mzahor/leany/formula-engine/src/visitor.ts:91:21)
    at InterpreterVisitor.visit (/Users/mzahor/leany/formula-engine/node_modules/chevrotain/src/parse/cst/cst_visitor.ts:61:32)
    at InterpreterVisitor.multiplicationExpression (/Users/mzahor/leany/formula-engine/src/visitor.ts:71:25)
    at InterpreterVisitor.visit (/Users/mzahor/leany/formula-engine/node_modules/chevrotain/src/parse/cst/cst_visitor.ts:61:32)
    at InterpreterVisitor.additionExpression (/Users/mzahor/leany/formula-engine/src/visitor.ts:52:25)
    at InterpreterVisitor.visit (/Users/mzahor/leany/formula-engine/node_modules/chevrotain/src/parse/cst/cst_visitor.ts:61:32)
    at InterpreterVisitor.comparisonExpression (/Users/mzahor/leany/formula-engine/src/visitor.ts:27:23)"
`;

exports[`Engine invalid syntax should throw LexerError 1`] = `"Tokenization error"`;

exports[`Engine invalid syntax should throw ParsingError 1`] = `"Parsing error"`;
